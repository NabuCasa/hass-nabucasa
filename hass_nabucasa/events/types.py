"""Event types for cloud system."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import TYPE_CHECKING, Literal

if TYPE_CHECKING:
    from hass_nabucasa.iot_base import DisconnectReason


def _timestamp_factory() -> float:
    """Generate a timestamp for the current time."""
    return datetime.now(UTC).timestamp()


@dataclass(kw_only=True)
class _CloudEventBase:
    """Base class for all cloud events."""

    timestamp: float = field(default_factory=_timestamp_factory)


@dataclass(kw_only=True)
class RelayerConnectedEvent(_CloudEventBase):
    """Relayer connected event."""

    type: Literal["relayer_connected"] = "relayer_connected"


@dataclass(kw_only=True)
class RelayerDisconnectedEvent(_CloudEventBase):
    """Relayer disconnected event."""

    type: Literal["relayer_disconnected"] = "relayer_disconnected"
    reason: DisconnectReason | None = None


CloudEvent = RelayerConnectedEvent | RelayerDisconnectedEvent
type CloudEventType = Literal["relayer_connected", "relayer_disconnected"]
