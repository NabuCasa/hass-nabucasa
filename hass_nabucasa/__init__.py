"""Component to integrate the Home Assistant cloud."""
from __future__ import annotations

import asyncio
from datetime import datetime, timedelta
import json
import logging
from pathlib import Path
from typing import Awaitable, Callable, Coroutine, List

import aiohttp
import async_timeout
from atomicwrites import atomic_write
from jose import jwt

from .auth import CloudError, CognitoAuth
from .client import CloudClient
from .cloudhooks import Cloudhooks
from .const import CONFIG_DIR, MODE_DEV, SERVERS, STATE_CONNECTED
from .google_report_state import GoogleReportState
from .iot import CloudIoT
from .remote import RemoteUI
from .utils import UTC, gather_callbacks, parse_date, utcnow
from .voice import Voice

_LOGGER = logging.getLogger(__name__)


class Cloud:
    """Store the configuration of the cloud connection."""

    def __init__(
        self,
        client: CloudClient,
        mode: str,
        cognito_client_id=None,
        user_pool_id=None,
        region=None,
        relayer=None,
        google_actions_report_state_url=None,
        subscription_info_url=None,
        cloudhook_create_url=None,
        remote_api_url=None,
        alexa_access_token_url=None,
        account_link_url=None,
        voice_api_url=None,
        acme_directory_server=None,
        thingtalk_url=None,
    ):
        """Create an instance of Cloud."""
        self._on_initialized: List[Callable[[], Awaitable[None]]] = []
        self._on_start: List[Callable[[], Awaitable[None]]] = []
        self._on_stop: List[Callable[[], Awaitable[None]]] = []
        self.mode = mode
        self.client = client
        self.id_token = None
        self.access_token = None
        self.refresh_token = None
        self.started = None
        self.iot = CloudIoT(self)
        self.google_report_state = GoogleReportState(self)
        self.cloudhooks = Cloudhooks(self)
        self.remote = RemoteUI(self)
        self.auth = CognitoAuth(self)
        self.voice = Voice(self)

        self._init_task = None

        # Set reference
        self.client.cloud = self

        if mode == MODE_DEV:
            self.cognito_client_id = cognito_client_id
            self.user_pool_id = user_pool_id
            self.region = region
            self.relayer = relayer
            self.google_actions_report_state_url = google_actions_report_state_url
            self.subscription_info_url = subscription_info_url
            self.cloudhook_create_url = cloudhook_create_url
            self.remote_api_url = remote_api_url
            self.alexa_access_token_url = alexa_access_token_url
            self.acme_directory_server = acme_directory_server
            self.account_link_url = account_link_url
            self.voice_api_url = voice_api_url
            self.thingtalk_url = thingtalk_url
            return

        info = SERVERS[mode]

        self.cognito_client_id = info["cognito_client_id"]
        self.user_pool_id = info["user_pool_id"]
        self.region = info["region"]
        self.relayer = info["relayer"]
        self.google_actions_report_state_url = info["google_actions_report_state_url"]
        self.subscription_info_url = info["subscription_info_url"]
        self.cloudhook_create_url = info["cloudhook_create_url"]
        self.remote_api_url = info["remote_api_url"]
        self.alexa_access_token_url = info["alexa_access_token_url"]
        self.account_link_url = info["account_link_url"]
        self.voice_api_url = info["voice_api_url"]
        self.acme_directory_server = info["acme_directory_server"]
        self.thingtalk_url = info["thingtalk_url"]

    @property
    def is_logged_in(self) -> bool:
        """Get if cloud is logged in."""
        return self.id_token is not None

    @property
    def is_connected(self) -> bool:
        """Return True if we are connected."""
        return self.iot.state == STATE_CONNECTED

    @property
    def websession(self) -> aiohttp.ClientSession:
        """Return websession for connections."""
        return self.client.websession

    @property
    def subscription_expired(self) -> bool:
        """Return a boolean if the subscription has expired."""
        return utcnow() > self.expiration_date + timedelta(days=7)

    @property
    def expiration_date(self) -> datetime:
        """Return the subscription expiration as a UTC datetime object."""
        return datetime.combine(
            parse_date(self.claims["custom:sub-exp"]), datetime.min.time()
        ).replace(tzinfo=UTC)

    @property
    def username(self) -> str:
        """Return the subscription username."""
        return self.claims["cognito:username"]

    @property
    def claims(self):
        """Return the claims from the id token."""
        return self._decode_claims(self.id_token)

    @property
    def user_info_path(self) -> Path:
        """Get path to the stored auth."""
        return self.path(f"{self.mode}_auth.json")

    async def update_token(
        self, id_token: str, access_token: str, refresh_token: str | None = None
    ) -> None:
        """Update the id and access token."""
        self.id_token = id_token
        self.access_token = access_token
        if refresh_token is not None:
            self.refresh_token = refresh_token

        await self.run_executor(self._write_user_info)

        if self.started is None:
            return

        if not self.started and not self.subscription_expired:
            self.started = True
            self.run_task(self._start())

        elif self.started and self.subscription_expired:
            self.started = False
            await self.stop()

    def register_on_initialized(self, on_initialized_cb: Callable[[], Awaitable[None]]):
        """Register an async on_initialized callback.

        on_initialized callbacks are called after all on_start callbacks.
        """
        self._on_initialized.append(on_initialized_cb)

    def register_on_start(self, on_start_cb: Callable[[], Awaitable[None]]):
        """Register an async on_start callback."""
        self._on_start.append(on_start_cb)

    def register_on_stop(self, on_stop_cb: Callable[[], Awaitable[None]]):
        """Register an async on_stop callback."""
        self._on_stop.append(on_stop_cb)

    def path(self, *parts) -> Path:
        """Get config path inside cloud dir.

        Async friendly.
        """
        return Path(self.client.base_path, CONFIG_DIR, *parts)

    def run_task(self, coro: Coroutine) -> Coroutine:
        """Schedule a task.

        Return a coroutine.
        """
        return self.client.loop.create_task(coro)

    def run_executor(self, callback: Callable, *args) -> asyncio.Future:
        """Run function inside executore.

        Return a awaitable object.
        """
        return self.client.loop.run_in_executor(None, callback, *args)

    async def login(self, email: str, password: str) -> None:
        """Log a user in."""
        async with async_timeout.timeout(30):
            await self.auth.async_login(email, password)

    async def logout(self) -> None:
        """Close connection and remove all credentials."""
        self.id_token = None
        self.access_token = None
        self.refresh_token = None

        self.started = False
        await self.stop()

        # Cleanup auth data
        if self.user_info_path.exists():
            await self.run_executor(self.user_info_path.unlink)

        await self.client.logout_cleanups()

    def _write_user_info(self) -> None:
        """Write user info to a file."""
        base_path = self.path()
        if not base_path.exists():
            base_path.mkdir()

        with atomic_write(self.user_info_path, overwrite=True) as fp:
            fp.write(
                json.dumps(
                    {
                        "id_token": self.id_token,
                        "access_token": self.access_token,
                        "refresh_token": self.refresh_token,
                    },
                    indent=4,
                )
            )
        self.user_info_path.chmod(0o600)

    async def initialize(self):
        """Initialize the cloud component (load auth and maybe start)."""

        def load_config():
            """Load config."""
            # Ensure config dir exists
            base_path = self.path()
            if not base_path.exists():
                base_path.mkdir()

            if not self.user_info_path.exists():
                return None
            try:
                return json.loads(self.user_info_path.read_text(encoding="utf-8"))
            except (ValueError, OSError) as err:
                path = self.user_info_path.relative_to(self.client.base_path)
                self.client.user_message(
                    "load_auth_data",
                    "Home Assistant Cloud error",
                    f"Unable to load authentication from {path}. [Please login again](/config/cloud)",
                )
                _LOGGER.warning(
                    "Error loading cloud authentication info from %s: %s", path, err
                )
                return None

        info = await self.run_executor(load_config)
        if info is None:
            # No previous token data
            self.started = False
            return

        self.id_token = info["id_token"]
        self.access_token = info["access_token"]
        self.refresh_token = info["refresh_token"]

        self._init_task = self.run_task(self._finish_initialize())

    async def _finish_initialize(self):
        """Finish initializing the cloud component (load auth and maybe start)."""
        try:
            await self.auth.async_check_token()
        except CloudError:
            _LOGGER.debug("Failed to check cloud token", exc_info=True)

        if self.subscription_expired:
            self.started = False
            return

        self.started = True
        await self._start()
        await gather_callbacks(_LOGGER, "on_initialized", self._on_initialized)
        self._init_task = None

    async def _start(self):
        """Start the cloud component."""
        await self.client.cloud_started()
        await gather_callbacks(_LOGGER, "on_start", self._on_start)

    async def stop(self):
        """Stop the cloud component."""
        if self._init_task:
            self._init_task.cancel()
            self._init_task = None

        await self.client.cloud_stopped()
        await gather_callbacks(_LOGGER, "on_stop", self._on_stop)

    @staticmethod
    def _decode_claims(token):
        """Decode the claims in a token."""
        return jwt.get_unverified_claims(token)
