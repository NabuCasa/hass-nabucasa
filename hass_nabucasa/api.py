"""Define the API base class."""

from __future__ import annotations

from abc import ABC, abstractmethod
import contextlib
from json import JSONDecodeError
import logging
from typing import TYPE_CHECKING, Any, final

from aiohttp import (
    ClientError,
    ClientResponse,
    ClientResponseError,
    ClientTimeout,
    hdrs,
)

from .auth import CloudError

if TYPE_CHECKING:
    from . import Cloud, _ClientT

_LOGGER = logging.getLogger(__name__)


class CloudApiError(CloudError):
    """Exception raised when handling cloud API."""

    def __init__(
        self,
        context: str | Exception,
        *,
        orig_exc: Exception | None = None,
    ) -> None:
        """Initialize."""
        super().__init__(context)
        self.orig_exc = orig_exc


class CloudApiCodedError(CloudApiError):
    """Exception raised when handling cloud API."""

    def __init__(self, context: str | Exception, *, code: str) -> None:
        """Initialize."""
        super().__init__(context)
        self.code = code


class CloudApiTimeoutError(CloudApiError):
    """Exception raised when handling cloud API times out."""


class CloudApiClientError(CloudApiError):
    """Exception raised when handling cloud API client error."""


class CloudApiNonRetryableError(CloudApiCodedError):
    """Exception raised when handling cloud API non-retryable error."""


class ApiBase(ABC):
    """Class to help communicate with the cloud API."""

    def __init__(self, cloud: Cloud[_ClientT]) -> None:
        """Initialize the API base."""
        self._cloud = cloud

    @property
    @abstractmethod
    def hostname(self) -> str:
        """Get the hostname."""

    @property
    def non_retryable_error_codes(self) -> set[str]:
        """Get the non-retryable error codes."""
        return set()

    @property
    @final
    def _non_retryable_error_codes(self) -> set[str]:
        """Get the non-retryable error codes."""
        return {"NC-CE-02", "NC-CE-03"}.union(self.non_retryable_error_codes)

    def _do_log_response(
        self,
        resp: ClientResponse,
        data: list[Any] | dict[Any, Any] | str | None = None,
    ) -> None:
        """Log the response."""
        isok = resp.status < 400
        target = resp.url.path if resp.url.host == self.hostname else ""
        _LOGGER.debug(
            "Response for %s from %s%s (%s) %s",
            resp.method,
            resp.url.host,
            target,
            resp.status,
            data["message"]
            if not isok and isinstance(data, dict) and "message" in data
            else "",
        )

    async def _call_raw_api(
        self,
        *,
        url: str,
        method: str,
        client_timeout: ClientTimeout,
        headers: dict[str, Any],
        jsondata: dict[str, Any] | None = None,
        data: Any | None = None,
    ) -> ClientResponse:
        """Call raw API."""
        try:
            resp = await self._cloud.websession.request(
                method=method,
                url=url,
                timeout=client_timeout,
                headers=headers,
                json=jsondata,
                data=data,
            )
        except TimeoutError as err:
            raise CloudApiTimeoutError(
                "Timeout reached while calling API",
                orig_exc=err,
            ) from err
        except ClientResponseError as err:
            raise CloudApiClientError(
                f"Failed to fetch: ({err.status}) {err.message}",
                orig_exc=err,
            ) from err
        except ClientError as err:
            raise CloudApiClientError(f"Failed to fetch: {err}", orig_exc=err) from err
        except Exception as err:
            raise CloudApiError(
                f"Unexpected error while calling API: {err}",
                orig_exc=err,
            ) from err

        return resp

    async def _call_cloud_api(
        self,
        *,
        path: str,
        method: str = "GET",
        client_timeout: ClientTimeout | None = None,
        jsondata: dict[str, Any] | None = None,
        headers: dict[str, Any] | None = None,
        skip_token_check: bool = False,
    ) -> Any:
        """Call cloud API."""
        data: dict[str, Any] | list[Any] | str | None = None
        if not skip_token_check:
            await self._cloud.auth.async_check_token()
        if TYPE_CHECKING:
            assert self._cloud.id_token is not None

        resp = await self._call_raw_api(
            method=method,
            url=f"https://{self.hostname}{path}",
            client_timeout=client_timeout or ClientTimeout(total=10),
            headers={
                hdrs.ACCEPT: "application/json",
                hdrs.AUTHORIZATION: self._cloud.id_token,
                hdrs.CONTENT_TYPE: "application/json",
                hdrs.USER_AGENT: self._cloud.client.client_name,
                **(headers or {}),
            },
            jsondata=jsondata,
        )

        if resp.status < 500:
            with contextlib.suppress(JSONDecodeError):
                data = await resp.json()

        self._do_log_response(resp, data)

        if data is None and resp.method.upper() != "DELETE":
            raise CloudApiError("Failed to parse API response") from None

        if (
            resp.status == 400
            and isinstance(data, dict)
            and (message := data.get("message"))
            and (code := message.split(" ")[0])
            and code in self._non_retryable_error_codes
        ):
            raise CloudApiNonRetryableError(message, code=code) from None

        try:
            resp.raise_for_status()
        except ClientResponseError as err:
            raise CloudApiError(
                f"Failed to fetch: ({err.status}) {err.message}",
                orig_exc=err,
            ) from err
        return data
