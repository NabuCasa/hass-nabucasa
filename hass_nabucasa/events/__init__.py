"""Event system for cloud services."""

from .bus import CloudEventBus
from .types import CloudEvent, CloudEventType

__all__ = [
    "CloudEvent",
    "CloudEventBus",
    "CloudEventType",
]
