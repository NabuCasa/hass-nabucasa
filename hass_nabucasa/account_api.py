"""Manage account API."""

from __future__ import annotations

from typing import TYPE_CHECKING, TypedDict

from .api import ApiBase, CloudApiError


class AccountApiError(CloudApiError):
    """Exception raised when handling account API."""


class AccountServicesDetails(TypedDict):
    """Details of the services."""

    alexa: AccountServiceDetails
    google_home: AccountServiceDetails
    remote_access: AccountServiceDetails
    stt: AccountServiceDetails
    storage: AccountStorageServiceDetails
    tts: AccountServiceDetails
    webhooks: AccountServiceDetails
    webrtc: AccountServiceDetails


class AccountServiceDetails(TypedDict):
    """Details of a service."""

    available: bool


class AccountStorageServiceDetails(AccountServiceDetails):
    """Details of a service."""

    limit_bytes: int


class AccountApi(ApiBase):
    """Class to help communicate with the instance API."""

    @property
    def hostname(self) -> str:
        """Get the hostname."""
        if TYPE_CHECKING:
            assert self._cloud.servicehandlers_server is not None
        return self._cloud.servicehandlers_server

    async def services(self) -> AccountServicesDetails:
        """Get the services details."""
        try:
            details: AccountServicesDetails = await self._call_cloud_api(
                path="/account/services",
            )
        except CloudApiError as err:
            raise AccountApiError(err, orig_exc=err) from err
        return details
