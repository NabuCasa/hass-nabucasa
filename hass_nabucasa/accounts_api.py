"""Manage accounts API."""

from __future__ import annotations

from typing import TYPE_CHECKING

from .api import ApiBase, CloudApiError, api_exception_handler


class AccountsApiError(CloudApiError):
    """Exception raised when handling account API."""


class AccountsApi(ApiBase):
    """Class to help communicate with the accounts API."""

    @property
    def hostname(self) -> str:
        """Get the hostname."""
        if TYPE_CHECKING:
            assert self._cloud.accounts_server is not None
        return self._cloud.accounts_server

    @api_exception_handler(AccountsApiError)
    async def instance_resolve_dns_cname(self, *, hostname: str) -> list[str]:
        """Resolve DNS CNAME."""
        details: list[str] = await self._call_cloud_api(
            method="POST",
            path="/instance/resolve_dns_cname",
            jsondata={"hostname": hostname},
        )
        return details
