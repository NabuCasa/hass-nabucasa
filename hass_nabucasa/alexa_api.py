"""This module provides Alexa API functionalities."""

from __future__ import annotations

from typing import TYPE_CHECKING, TypedDict

from .api import ApiBase, CloudApiError, api_exception_handler


class AlexaApiError(CloudApiError):
    """Exception raised when handling Alexa API."""


class AlexaAccessTokenDetails(TypedDict):
    """Alexa access token details from Alexa API."""

    access_token: str
    expires_in: int
    event_endpoint: str


class AlexaApi(ApiBase):
    """Class to help communicate with the Alexa API."""

    @property
    def hostname(self) -> str:
        """Get the hostname."""
        if TYPE_CHECKING:
            assert self._cloud.servicehandlers_server is not None
        return self._cloud.servicehandlers_server

    @api_exception_handler(AlexaApiError)
    async def access_token(self) -> AlexaAccessTokenDetails:
        """Get the Alexa API access token."""
        details: AlexaAccessTokenDetails = await self._call_cloud_api(
            method="POST", path="/alexa/access_token"
        )
        return details
