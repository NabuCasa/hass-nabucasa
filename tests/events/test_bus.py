"""Tests for event bus."""

from unittest.mock import AsyncMock

import pytest
from syrupy import SnapshotAssertion

from hass_nabucasa.events import CloudEventBus
from hass_nabucasa.events.types import (
    RelayerConnectedEvent,
    RelayerDisconnectedEvent,
)
from tests.common import extract_log_messages


@pytest.mark.asyncio
async def test_subscribe_and_publish(
    caplog: pytest.LogCaptureFixture,
    snapshot: SnapshotAssertion,
):
    """Test basic subscription and publishing."""
    subscriber = AsyncMock()
    event_bus = CloudEventBus()

    unsubscribe = event_bus.subscribe(
        event_type="relayer_connected", handler=subscriber
    )

    await event_bus.publish(event=RelayerConnectedEvent())

    assert len(subscriber.call_args_list) == 1

    unsubscribe()

    assert subscriber.call_args_list == snapshot
    assert extract_log_messages(caplog) == snapshot


@pytest.mark.asyncio
async def test_publish_without_data(
    caplog: pytest.LogCaptureFixture,
    snapshot: SnapshotAssertion,
):
    """Test publishing without extra fields."""
    subscriber = AsyncMock()
    event_bus = CloudEventBus()

    event_bus.subscribe(event_type="relayer_connected", handler=subscriber)

    event = RelayerConnectedEvent()
    await event_bus.publish(event=event)

    assert len(subscriber.call_args_list) == 1
    assert subscriber.call_args_list[0][0][0].type == "relayer_connected"
    assert extract_log_messages(caplog) == snapshot


@pytest.mark.asyncio
async def test_publish_without_subscribers(
    caplog: pytest.LogCaptureFixture,
    snapshot: SnapshotAssertion,
):
    """Test publishing without any subscribers doesn't raise."""
    event_bus = CloudEventBus()

    event = RelayerConnectedEvent()
    await event_bus.publish(event=event)

    assert extract_log_messages(caplog) == snapshot


@pytest.mark.asyncio
async def test_unsubscribe(
    caplog: pytest.LogCaptureFixture,
    snapshot: SnapshotAssertion,
):
    """Test unsubscribing from events."""
    subscriber = AsyncMock()
    event_bus = CloudEventBus()

    unsubscribe = event_bus.subscribe(
        event_type="relayer_connected", handler=subscriber
    )

    await event_bus.publish(event=RelayerConnectedEvent())
    assert len(subscriber.call_args_list) == 1

    unsubscribe()

    await event_bus.publish(event=RelayerConnectedEvent())
    assert len(subscriber.call_args_list) == 1

    assert extract_log_messages(caplog) == snapshot


@pytest.mark.asyncio
async def test_unsubscribe_multiple_times(
    caplog: pytest.LogCaptureFixture,
    snapshot: SnapshotAssertion,
):
    """Test that unsubscribing multiple times doesn't raise."""
    subscriber = AsyncMock()
    event_bus = CloudEventBus()

    unsubscribe = event_bus.subscribe(
        event_type="relayer_connected", handler=subscriber
    )

    unsubscribe()
    unsubscribe()

    await event_bus.publish(event=RelayerConnectedEvent())
    assert len(subscriber.call_args_list) == 0

    assert extract_log_messages(caplog) == snapshot


@pytest.mark.asyncio
async def test_error_handling(
    caplog: pytest.LogCaptureFixture,
    snapshot: SnapshotAssertion,
):
    """Test that errors in handlers don't stop other handlers."""
    failing_subscriber = AsyncMock(side_effect=ValueError("Test error"))
    working_subscriber = AsyncMock()
    event_bus = CloudEventBus()

    event_bus.subscribe(event_type="relayer_connected", handler=failing_subscriber)
    event_bus.subscribe(event_type="relayer_connected", handler=working_subscriber)

    await event_bus.publish(event=RelayerConnectedEvent())

    assert len(failing_subscriber.call_args_list) == 1
    assert len(working_subscriber.call_args_list) == 1
    assert extract_log_messages(caplog) == snapshot


@pytest.mark.asyncio
async def test_multiple_event_types(
    caplog: pytest.LogCaptureFixture,
    snapshot: SnapshotAssertion,
):
    """Test subscribing to different event types."""
    connected_subscriber = AsyncMock()
    disconnected_subscriber = AsyncMock()
    event_bus = CloudEventBus()

    event_bus.subscribe(event_type="relayer_connected", handler=connected_subscriber)
    event_bus.subscribe(
        event_type="relayer_disconnected", handler=disconnected_subscriber
    )

    await event_bus.publish(event=RelayerConnectedEvent())
    await event_bus.publish(event=RelayerDisconnectedEvent())

    assert len(connected_subscriber.call_args_list) == 1
    assert len(disconnected_subscriber.call_args_list) == 1
    assert extract_log_messages(caplog) == snapshot


@pytest.mark.asyncio
async def test_multiple_subscribers(
    caplog: pytest.LogCaptureFixture,
    snapshot: SnapshotAssertion,
):
    """Test multiple subscribers to the same event."""
    subscriber1 = AsyncMock()
    subscriber2 = AsyncMock()
    subscriber3 = AsyncMock()
    event_bus = CloudEventBus()

    event_bus.subscribe(event_type="relayer_connected", handler=subscriber1)
    event_bus.subscribe(event_type="relayer_connected", handler=subscriber2)
    event_bus.subscribe(event_type="relayer_connected", handler=subscriber3)

    await event_bus.publish(event=RelayerConnectedEvent())

    assert len(subscriber1.call_args_list) == 1
    assert len(subscriber2.call_args_list) == 1
    assert len(subscriber3.call_args_list) == 1
    assert extract_log_messages(caplog) == snapshot


@pytest.mark.asyncio
async def test_event_timestamp(
    caplog: pytest.LogCaptureFixture,
    snapshot: SnapshotAssertion,
):
    """Test that events have timestamps."""
    subscriber = AsyncMock()
    event_bus = CloudEventBus()

    event_bus.subscribe(event_type="relayer_connected", handler=subscriber)

    await event_bus.publish(event=RelayerConnectedEvent())

    assert len(subscriber.call_args_list) == 1
    assert subscriber.call_args_list[0][0][0].timestamp is not None
    assert extract_log_messages(caplog) == snapshot


@pytest.mark.asyncio
async def test_subscribe_to_multiple_event_types(
    caplog: pytest.LogCaptureFixture,
    snapshot: SnapshotAssertion,
):
    """Test subscribing to multiple event types with a single handler."""
    subscriber = AsyncMock()
    event_bus = CloudEventBus()

    unsubscribe = event_bus.subscribe(
        event_type=["relayer_connected", "relayer_disconnected"], handler=subscriber
    )

    await event_bus.publish(event=RelayerConnectedEvent())
    await event_bus.publish(event=RelayerDisconnectedEvent())

    assert len(subscriber.call_args_list) == 2
    assert subscriber.call_args_list[0][0][0].type == "relayer_connected"
    assert subscriber.call_args_list[1][0][0].type == "relayer_disconnected"

    unsubscribe()

    await event_bus.publish(event=RelayerConnectedEvent())
    assert len(subscriber.call_args_list) == 2

    assert extract_log_messages(caplog) == snapshot
