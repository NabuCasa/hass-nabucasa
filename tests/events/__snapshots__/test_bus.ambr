# serializer version: 1
# name: test_error_handling
  '''
  [DEBUG] hass_nabucasa.events.bus: Publish relayer_connected to 2 subscribers
  [WARNING] hass_nabucasa.events.bus: Error in event handler AsyncMock for event relayer_connected
  '''
# ---
# name: test_event_timestamp
  '[DEBUG] hass_nabucasa.events.bus: Publish relayer_connected to 1 subscribers'
# ---
# name: test_multiple_event_types
  '''
  [DEBUG] hass_nabucasa.events.bus: Publish relayer_connected to 1 subscribers
  [DEBUG] hass_nabucasa.events.bus: Publish relayer_disconnected to 1 subscribers
  '''
# ---
# name: test_multiple_subscribers
  '[DEBUG] hass_nabucasa.events.bus: Publish relayer_connected to 3 subscribers'
# ---
# name: test_publish_without_data
  '[DEBUG] hass_nabucasa.events.bus: Publish relayer_connected to 1 subscribers'
# ---
# name: test_publish_without_subscribers
  ''
# ---
# name: test_subscribe_and_publish
  _CallList([
    _Call(
      tuple(
        RelayerConnectedEvent(timestamp=1537185600, type='relayer_connected'),
      ),
      dict({
      }),
    ),
  ])
# ---
# name: test_subscribe_and_publish.1
  '[DEBUG] hass_nabucasa.events.bus: Publish relayer_connected to 1 subscribers'
# ---
# name: test_subscribe_to_multiple_event_types
  '''
  [DEBUG] hass_nabucasa.events.bus: Publish relayer_connected to 1 subscribers
  [DEBUG] hass_nabucasa.events.bus: Publish relayer_disconnected to 1 subscribers
  [DEBUG] hass_nabucasa.events.bus: No subscribers for event relayer_connected
  '''
# ---
# name: test_unsubscribe
  '''
  [DEBUG] hass_nabucasa.events.bus: Publish relayer_connected to 1 subscribers
  [DEBUG] hass_nabucasa.events.bus: No subscribers for event relayer_connected
  '''
# ---
# name: test_unsubscribe_multiple_times
  '[DEBUG] hass_nabucasa.events.bus: No subscribers for event relayer_connected'
# ---
